nsis/src/nsis/file/NsisExecutable.java

```diff
@@ -35,49 +42,117 @@ public NsisExecutable() {
 		this.headerOffset = findHeaderOffset();
-		initScriptHeader(bp);
+		initScriptHeader();
+		if ((this.scriptHeader.compressedHeaderSize & 0x80000000) != 0) { // Check
+																			// if
+																			// MSB
+																			// is
+																			// set
+			decompressData();
+		}
 	}
 
 	private long findHeaderOffset() throws IOException, InvalidFormatException {
-		for (long headerOffset = 0; headerOffset
-				+ NsisConstants.NSIS_MAGIC.length <= reader
-						.length(); headerOffset++) {
+		for (long headerOffset = 0; headerOffset + NsisConstants.NSIS_SIGINFO.length
+				+ NsisConstants.NSIS_MAGIC.length <= reader.length(); headerOffset++) {
 			byte[] content = reader.readByteArray(headerOffset,
-					NsisConstants.NSIS_MAGIC.length);
-			if (Arrays.equals(NsisConstants.NSIS_MAGIC, content)) {
-				return headerOffset;
+					NsisConstants.NSIS_SIGINFO.length + NsisConstants.NSIS_MAGIC.length);
+			if (Arrays.equals(Bytes.concat(NsisConstants.NSIS_SIGINFO, NsisConstants.NSIS_MAGIC),
+					content)) {
+				return headerOffset - StructConverter.DWORD.getLength(); // Include
+																			// flags
+																			// in
+																			// header
 			}
 		}
 		throw new InvalidFormatException("Nsis magic not found.");
 	}
 
-	private void initScriptHeader(ByteProvider bp)
-			throws IOException, InvalidFormatException {
-		BinaryReader br = new BinaryReader(bp, /* isLittleEndian= */ true);
-		br.setPointerIndex(this.headerOffset);
-		this.scriptHeader = new NsisScriptHeader(br);
+	private void initScriptHeader() throws IOException, InvalidFormatException {
+		this.reader.setPointerIndex(this.headerOffset);
+		this.scriptHeader = new NsisScriptHeader(this.reader);
+	}
+
+	/**
+	 * Attempt to decompress the data from the reader. Supports LZMA algorithm. Will
+	 * eventually support Bzip2 and Zlib.
+	 * 
+	 * @throws IOException
+	 */
+	private void decompressData() throws IOException {
+		byte compressionByte = this.reader.readNextByte();
+		long compressedDataOffset;
+		int compressedDataLength;
+		byte[] compressedData;
+		if (isLZMA(compressionByte)) {
+			int dictionarySize = this.reader.readNextInt();
+			compressedDataOffset = this.headerOffset + NsisScriptHeader.getHeaderSize()
+					+ NsisConstants.COMPRESSION_LZMA_HEADER_LENGTH;
+			compressedDataLength = (this.scriptHeader.compressedHeaderSize ^ 0x80000000)
+					- NsisConstants.COMPRESSION_LZMA_HEADER_LENGTH; // Flip the
+																	// MSB to
+																	// get the
+																	// length
+			compressedData = this.reader.readByteArray(compressedDataOffset, compressedDataLength);
+			decompressLZMA(compressedData, compressionByte, dictionarySize);
+		} else if (isBzip2(compressionByte)) {
+			// TODO Bzip2 decompress
+			System.out.println("Decompress Bzip");
+		} else {
+			// TODO Zlib decompress
+			System.out.println("Decompress Zlib");
+		}
+		return;
+	}
+
+	private boolean isLZMA(byte significantByte) {
```

Comment:
What do you think of just doing this comparison where it is used instead of having another function?

Reply:
I thought the code would be more readable that way, but I think it is easily understandable that we are checking which type of compression it is if I include the comparison in the condition instead of having boolean functions